plugins/tekton/src/utils/pipeline-topology-utils.ts

```diff
@@ -0,0 +1,589 @@
+});
+
+export const getWhenStatus = (
+  status: ComputedStatus,
+): WhenStatus | undefined => {
+  switch (status) {
+    case ComputedStatus.Succeeded:
+    case ComputedStatus.Failed:
+      return WhenStatus.Met;
+    case ComputedStatus.Skipped:
+    case ComputedStatus['In Progress']:
+    case ComputedStatus.Idle:
+      return WhenStatus.Unmet;
+    default:
+      return undefined;
+  }
+};
+
+export const getTaskWhenStatus = (
+  task: PipelineTaskWithStatus,
+): WhenStatus | undefined => {
+  if (!task.when) {
+    return undefined;
+  }
+  return getWhenStatus(task.status?.reason as ComputedStatus);
+};
+
+const getDepsFromContextVariables = (task: PipelineTask) => {
+  const depsFromContextVariables: string[] = [];
+  if (task.params) {
+    task.params.forEach((p: PipelineTaskParam) => {
+      if (Array.isArray(p.value)) {
+        p.value.forEach(paramValue => {
+          depsFromContextVariables.push(
+            ...extractDepsFromContextVariables(paramValue),
+          );
+        });
+      } else {
+        depsFromContextVariables.push(
+          ...extractDepsFromContextVariables(p.value),
+        );
+      }
+    });
+  }
+  if (task?.when) {
+    task.when.forEach(({ input, values }) => {
+      depsFromContextVariables.push(...extractDepsFromContextVariables(input));
+      values.forEach((whenValue: string) => {
+        depsFromContextVariables.push(
+          ...extractDepsFromContextVariables(whenValue),
+        );
+      });
+    });
+  }
+  return depsFromContextVariables;
+};
+
+const getRunAfterTasks = (task: PipelineTask, dag: DAG, vertex: Vertex) => {
+  const runAfterTasks: string[] = [];
+  const depsFromContextVariables = getDepsFromContextVariables(task);
+
+  const dependancies = uniq([...vertex.dependancyNames]);
+  if (dependancies) {
+    dependancies.forEach(dep => {
+      const depObj = dag.vertices.get(dep) as Vertex;
+      if (
+        depObj.level - vertex.level <= 1 ||
+        vertex.data.runAfter?.includes(depObj.name)
+      ) {
+        runAfterTasks.push(dep);
+      }
+    });
+  }
+  if (depsFromContextVariables.length > 0) {
+    const v = depsFromContextVariables.map(d => {
+      return dag.vertices.get(d) as Vertex;
+    });
+    const minLevelDep = minBy(v, (d: Vertex) => d.level) as Vertex;
+    const nearestDeps = v.filter(v1 => v1.level === minLevelDep.level);
+    nearestDeps.forEach(nd => {
+      if (nd.level - vertex.level <= 1 || vertex.dependancyNames.length === 0) {
+        runAfterTasks.push(nd.name);
+      }
+    });
+  }
+  return runAfterTasks;
+};
+
+export const getGraphDataModel = (
+  pipelineRun: PipelineRunKind | undefined,
+  taskRuns: TaskRunKind[],
+): {
+  graph: GraphModel;
+  nodes: PipelineMixedNodeModel[];
+  edges: EdgeModel[];
+} | null => {
+  if (!pipelineRun) {
+    return null;
+  }
+
+  const plrTaskRuns = getPLRTaskRuns(taskRuns, pipelineRun?.metadata?.name);
+
+  const taskList = flatten(getPipelineTasks(pipelineRun, plrTaskRuns));
```

Comment:
you can directly append the status, we can avoid the extraneous code to make arrays and then flatten it at the end. (this probably needs change in ODC too).

```suggestion
  const taskList = appendPipelineRunStatus(pipelineRun, plrTaskRuns);
```